org.lflang/src/org/lflang/cli/Lfc.java

```diff
@@ -52,83 +56,98 @@ public Lfc() {
-         * @param isReq       Whether this option is required. If it is
-         *                    required but not specified a menu is shown.
-         * @param passOn      Whether to pass this option as a property
-         *                    to the code generator.
-         */
-        CLIOption(BuildParm parameter, String shorthand, boolean hasArg, boolean isReq, boolean passOn) {
-            this.option = new Option(shorthand, parameter.getKey(), hasArg, parameter.description);
-            option.setRequired(isReq);
-            this.passOn = passOn;
-        }
 
-        /**
-         * Create an Apache Commons CLI Options object and add all the options.
-         *
-         * @return Options object that includes all the options in this enum.
-         */
-        public static Options getOptions() {
-            Options opts = new Options();
-            Arrays.asList(CLIOption.values()).forEach(o -> opts.addOption(o.option));
-            return opts;
-        }
+    @Option(
+        names = "--build-type",
+        description = "The build type to use.")
+    private String buildType;
 
-        /**
-         * Return a list of options that are to be passed on to the code
-         * generator.
-         *
-         * @return List of options that must be passed on to the code gen stage.
-         */
-        public static List<Option> getPassedOptions() {
-            return Arrays.stream(CLIOption.values())
-                .filter(opt -> opt.passOn).map(opt -> opt.option)
-                .collect(Collectors.toList());
-        }
-    }
+    @Option(
+        names = {"-c", "--clean"},
+        arity = "0",
+        description = "Clean before building.")
+    private boolean clean;
+
+    @Option(
+        names = "--target-compiler",
+        description = "Target compiler to invoke.")
+    private String targetCompiler;
+
+    @Option(
+        names = "--external-runtime-path",
+        description = "Specify an external runtime library to be used by the"
+                    + " compiled binary.")
+    private String externalRuntimePath;
+
+    @Option(
+        names = {"-f", "--federated"},
+        arity = "0",
+        description = "Treat main reactor as federated.")
+    private boolean federated;
+
+    @Option(
+        names = "--logging",
+        description = "The logging level to use by the generated binary")
+    private String logging;
+
+    @Option(
+        names = {"-l", "--lint"},
+        arity = "0",
+        description = "Enable or disable linting of generated code.")
+    private boolean lint;
+
+    @Option(
+        names = {"-n", "--no-compile"},
+        arity = "0",
+        description = "Do not invoke target compiler.")
+    private boolean noCompile;
+
+    @Option(
+        names = {"-o", "--output-path"},
+        defaultValue = "",
+        fallbackValue = "",
+        description = "Specify the root output directory.")
+    private String outputPath;
```

Comment:
Also here we might want to leave it to picocli to check if the directory exists (if possible)

Reply:
Unfortunately Picocli doesn't do any validation beyond type conversion, so we'll still have to implement it ourselves. On the other hand, your comment gave me the idea of abstracting away redundant path parsing/validation code in `Lfc.java` and `Lff.java` into separate functions in `CliBase.java`, which made the code a lot cleaner.

Reply:
You can implement a custom validating converter very easily: https://picocli.info/#_custom_type_converters